gradio/queueing.py

```diff
@@ -345,81 +368,70 @@ async def clean_events(
                     if job.session_hash == session_hash or job._id == event_id:
                         job.alive = False
 
-        events_to_remove = []
-        for event in self.event_queue:
-            if event.session_hash == session_hash or event._id == event_id:
-                events_to_remove.append(event)
+        async with self.delete_lock:
+            events_to_remove: list[Event] = []
+            for event_queue in self.event_queue_per_concurrency_id.values():
+                for event in event_queue.queue:
+                    if event.session_hash == session_hash or event._id == event_id:
+                        events_to_remove.append(event)
 
-        for event in events_to_remove:
-            async with self.delete_lock:
-                self.event_queue.remove(event)
+            for event in events_to_remove:
+                self.event_queue_per_concurrency_id[event.concurrency_id].queue.remove(
+                    event
+                )
 
     async def notify_clients(self) -> None:
         """
         Notify clients about events statuses in the queue periodically.
         """
         while not self.stopped:
             await asyncio.sleep(self.update_intervals)
-            if self.event_queue:
-                await self.broadcast_estimations()
-
-    async def broadcast_estimations(self) -> None:
-        estimation = self.get_estimation()
-        # Send all messages concurrently
-        await asyncio.gather(
-            *[
-                self.send_estimation(event, estimation, rank)
-                for rank, event in enumerate(self.event_queue)
-            ]
-        )
-
-    async def send_estimation(
-        self, event: Event, estimation: Estimation, rank: int
-    ) -> Estimation:
-        """
-        Send estimation about ETA to the client.
+            if len(self) > 0:
+                for concurrency_id in self.event_queue_per_concurrency_id:
+                    self.broadcast_estimations(concurrency_id)
 
-        Parameters:
-            event:
-            estimation:
-            rank:
-        """
-        estimation.rank = rank
-
-        if self.avg_concurrent_process_time is not None:
-            estimation.rank_eta = (
-                estimation.rank * self.avg_concurrent_process_time
-                + self.avg_process_time
+    def broadcast_estimations(
+        self, concurrency_id: str, after: int | None = None
+    ) -> None:
+        wait_so_far = 0
```

Comment:
Can you explain what `wait_so_far` is measuring and why it can be `None`? Why is it added to `process_time_for_fn`?

Reply:
wait_so_far is the running estimated wait time. As we iterate through the events in the queue, we add the expected runtime of the each event It gets set to None if we encounter an event who's runtime we do not know (first time running this particular fn). In this case, this event and all events behind in queue will not recieve an estimate ETA.